File: day11.cc
#include <iostream>

int power_level (int x, int y, int serial)
{
	int p = (y*(x+10)+serial)*(x+10);
	p = ((int)(p/100)) % 10;
	return p-5;
}

int* build_grid (int serial, int n)
{
	int* grid = new int [n*n];
	for (int i=0; i<n; ++i)
		for (int j=0; j<n; ++j)
			grid[i+j*n] = power_level(i+1, j+1, serial);
	return grid;
}

int find_max (int* g, int g_size, int n, int& spot_x, int& spot_y)
{
	int max_p = 0;
	int x, y, i, j, p;
	for (x = 0; x < g_size-n; ++x)
		for (y = 0; y < g_size-n; ++y)
		{
			p = 0;
			for (i = 0; i < n; ++i)
				for (j = 0; j<n; ++j)
					p += g[(x+i) + (y+j) * g_size];
			if (p > max_p)
			{
				max_p = p;
				spot_x = x;
				spot_y = y;
			}
		}
	std::cout << "func: max for " << n << " is " << max_p << " at " << spot_x << " " << spot_y << std::endl;
	return max_p;
}

int find_all_max (int* g, int g_size, int& spot_x, int& spot_y)
{
	int p, max_p = 0, max_s = 0;
	int spot_xi, spot_yi;
	for (int s = 1; s < g_size; ++s)
	{
		p = find_max (g, g_size, s, spot_xi, spot_yi);
		if (p > max_p)
		{
			max_p = p;
			max_s = s;
			spot_x = spot_xi;
			spot_y = spot_yi;
		}
		std::cout << "max for size " << s << " is " << p << " at " << spot_xi+1 << ", " << spot_yi+1
				  << " max: (" << spot_x+1 << "," << spot_y+1 << "," << max_s << ")" << std::endl;
	}
	return 0;
}

int main (int argc, char** argv)
{
	int serial = atoi (argv[1]);
	std::cout << "serial is " << serial << std::endl;
	int* g = build_grid (serial, 300);
	int spot_x, spot_y, power;
	power = find_all_max (g, 300, spot_x, spot_y);
	return 0;
}

File: day11.py
def power_level (x, y, serial):
    rack_id = x+10
    p = (y*(x+10)+serial)*(x+10)
    p = (p // 100) % 10
    return p-5


def gen_grid (serial, n):
    grid = [[0] * n for i in range(n)]
    for x in range(n):
        for y in range(n):
            grid[x][y] = power_level(x+1, y+1, serial)
    return grid

def find_max (g, n):
    max_p = 0
    spot = None
    for x in range(len(g)-n+1):
        for y in range(len(g)-n+1):
            p = sum([sum([g[x+i][y+j] for i in range(n)]) for j in range(n)])
            if p > max_p:
                max_p = p
                spot = (x+1,y+1)
    return spot, max_p

# print (power_level(122,  79, 57))
# print (power_level(217, 196, 39))
# print (power_level(101, 153, 71))

g = gen_grid (18, 300)
find_max(g,3)

max_p = 0
max_size = 0
max_spot = 0
for i in range(1, 300):
    spot, p = find_max(g,i)
    if p > max_p:
        max_p = p
        max_spot = spot
        max_size = i
    print (i, "done")
print (max_spot, max_size, max_power)




